// @sap/ux-specification/dist/src/sync/v4/application.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../common/decoration/decorators");
exports.getSettingsPathForPage = (pageName) => {
    return `sap.ui5/routing/targets/${pageName[0]}/options/settings`;
};
exports.getSettingsPathForLRViewsPaths = (pageName) => {
    return exports.getSettingsPathForPage(pageName) + `/views/paths`;
};
exports.getSettingsPathForHeaderOP = (pageName) => {
    return exports.getSettingsPathForPage(pageName) + `/content/header`;
};
exports.getControlConfigurationPath = (pageName) => {
    return exports.getSettingsPathForPage(pageName) + '/controlConfiguration';
};
exports.getHeaderFacetSettingPath = (pageName, manifest, sectionId) => {
    return exports.getControlConfigurationPath(pageName) + `/@${"com.sap.vocabularies.UI.v1.HeaderFacets" /* HeaderFacets */}/facets/${sectionId}`;
};
exports.getTableSettingsPathForLR = (pageName) => {
    return exports.getControlConfigurationPath(pageName) + `/@${"com.sap.vocabularies.UI.v1.LineItem" /* LineItem */}/tableSettings`;
};
exports.getColumnsPathForLR = (pageName, manifest, sectionId, columnKey) => {
    return exports.getControlConfigurationPath(pageName) + `/@${"com.sap.vocabularies.UI.v1.LineItem" /* LineItem */}/columns/${columnKey}`;
};
exports.getColumnsPathForView = (pageName, manifest, targetAnnotationEncoded, columnKey, targetAnnotation) => {
    const annotationEncoded = targetAnnotation.replace(/\//gi, '??');
    return exports.getControlConfigurationPath(pageName) + `/${annotationEncoded}/columns/${columnKey}`;
};
exports.getActionsPathForLR = (pageName, manifest, sectionId, columnKey) => {
    return exports.getControlConfigurationPath(pageName) + `/@${"com.sap.vocabularies.UI.v1.LineItem" /* LineItem */}/actions/${columnKey}`;
};
exports.getActionsPathForView = (pageName, manifest, targetAnnotationEncoded, columnKey, targetAnnotation) => {
    const annotationEncoded = targetAnnotation.replace(/\//gi, '??');
    return exports.getControlConfigurationPath(pageName) + `/${annotationEncoded}/actions/${columnKey}`;
};
exports.getColumnsPathForOP = (pageName, manifest, sectionId, columnKey) => {
    return exports.getControlConfigurationPath(pageName) + `/${sectionId}/columns/${columnKey}`;
};
exports.getActionsPathForOP = (pageName, manifest, sectionId, columnKey) => {
    return exports.getControlConfigurationPath(pageName) + `/${sectionId}/actions/${columnKey}`;
};
exports.getTableSettingsPathForOP = (pageName, manifest, sectionId) => {
    return exports.getControlConfigurationPath(pageName) + `/${sectionId}/tableSettings`;
};
exports.getCustomSectionPath = (pageName, manifest, sectionId, pathPart) => {
    let path = exports.getSettingsPathForPage(pageName) + `/content/body/sections/${sectionId}`;
    if (pathPart) {
        path += '/' + pathPart;
    }
    return path;
};
class ApplicationV4 {
}
__decorate([
    decorators_1.childId('settings')
], ApplicationV4.prototype, "settings", void 0);
exports.ApplicationV4 = ApplicationV4;
class AppSettings {
}
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: () => 'sap.app'
        }
    })
], AppSettings.prototype, "title", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: () => 'sap.app'
        }
    })
], AppSettings.prototype, "description", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: () => 'sap.ui5/routing/config'
        }
    }),
    decorators_1.validity({
        since: '1.78.0'
    })
], AppSettings.prototype, "flexibleColumnLayout", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: () => 'sap.ui5'
        }
    }),
    decorators_1.validity({
        since: '1.56.0'
    })
], AppSettings.prototype, "flexEnabled", void 0);
exports.AppSettings = AppSettings;
