// @sap/ux-specification/dist/src/specification/v2/controls/Card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ChartCardType;
(function (ChartCardType) {
    ChartCardType["cardBubble"] = "cardBubble";
    ChartCardType["cardchartsline"] = "cardchartsline";
    ChartCardType["cardchartsdonut"] = "cardchartsdonut";
})(ChartCardType = exports.ChartCardType || (exports.ChartCardType = {}));
var CardTemplateType;
(function (CardTemplateType) {
    CardTemplateType["analytical"] = "sap.ovp.cards.charts.analytical";
    CardTemplateType["analyticalv4"] = "sap.ovp.cards.v4.charts.analytical";
    CardTemplateType["list"] = "sap.ovp.cards.list";
    CardTemplateType["listv4"] = "sap.ovp.cards.v4.list";
    CardTemplateType["linklist"] = "sap.ovp.cards.linklist";
    CardTemplateType["table"] = "sap.ovp.cards.table";
    CardTemplateType["tablev4"] = "sap.ovp.cards.v4.table";
    CardTemplateType["stack"] = "sap.ovp.cards.stack";
})(CardTemplateType = exports.CardTemplateType || (exports.CardTemplateType = {}));
exports.cardTemplateTypeMap = {
    'sap.ovp.cards.charts.analytical': 'AnalyticalCard',
    'sap.ovp.cards.v4.charts.analytical': 'AnalyticalCard',
    'sap.ovp.cards.list': 'ListCard',
    'sap.ovp.cards.v4.list': 'ListCard',
    'sap.ovp.cards.linklist': 'LinklistCard',
    'sap.ovp.cards.table': 'TableCard',
    'sap.ovp.cards.v4.table': 'TableCard',
    'sap.ovp.cards.stack': 'StackCard'
};
var CardSettingsType;
(function (CardSettingsType) {
    CardSettingsType["analyticalCardSettings"] = "analyticalCardSettings";
    CardSettingsType["analyticalCardSettingsv4"] = "analyticalCardSettingsv4";
    CardSettingsType["listCardSettings"] = "listCardSettings";
    CardSettingsType["listCardSettingsv4"] = "listCardSettingsv4";
    CardSettingsType["stackCardSettings"] = "stackCardSettings";
    CardSettingsType["linkListCardSettings"] = "linkListCardSettings";
    CardSettingsType["tableCardSettings"] = "tableCardSettings";
    CardSettingsType["tableCardSettingsv4"] = "tableCardSettingsv4";
})(CardSettingsType = exports.CardSettingsType || (exports.CardSettingsType = {}));
var MeasureAggregateValues;
(function (MeasureAggregateValues) {
    MeasureAggregateValues["average"] = "average";
    MeasureAggregateValues["max"] = "max";
    MeasureAggregateValues["min"] = "min";
    MeasureAggregateValues["sum"] = "sum";
    MeasureAggregateValues["count"] = "$count";
})(MeasureAggregateValues = exports.MeasureAggregateValues || (exports.MeasureAggregateValues = {}));
var ListFlavorType;
(function (ListFlavorType) {
    ListFlavorType["standard"] = "standard";
    ListFlavorType["bar"] = "bar";
    ListFlavorType["carousel"] = "carousel";
})(ListFlavorType = exports.ListFlavorType || (exports.ListFlavorType = {}));
var ListTypeType;
(function (ListTypeType) {
    ListTypeType["extended"] = "extended";
    ListTypeType["condensed"] = "condensed";
})(ListTypeType = exports.ListTypeType || (exports.ListTypeType = {}));
var SortOrderType;
(function (SortOrderType) {
    SortOrderType["ascending"] = "ascending";
    SortOrderType["descending"] = "descending";
})(SortOrderType = exports.SortOrderType || (exports.SortOrderType = {}));
var LinkListFlavorType;
(function (LinkListFlavorType) {
    LinkListFlavorType["standard"] = "standard";
    LinkListFlavorType["carousel"] = "carousel";
})(LinkListFlavorType = exports.LinkListFlavorType || (exports.LinkListFlavorType = {}));
