// @sap/ux-specification/dist/src/specification/v2/controls/FilterBar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DateRangeType;
(function (DateRangeType) {
    DateRangeType["DATERANGE"] = "DATERANGE";
    DateRangeType["DATE"] = "DATE";
    DateRangeType["FROM"] = "FROM";
    DateRangeType["TO"] = "TO";
    DateRangeType["DAYS"] = "DAYS";
    DateRangeType["LASTDAYS"] = "LASTDAYS";
    DateRangeType["LASTWEEKS"] = "LASTWEEKS";
    DateRangeType["WEEK"] = "WEEK";
    DateRangeType["LASTMONTHS"] = "LASTMONTHS";
    DateRangeType["MONTH"] = "MONTH";
    DateRangeType["QUARTER"] = "QUARTER";
    DateRangeType["LASTQUARTERS"] = "LASTQUARTERS";
    DateRangeType["LASTYEARS"] = "LASTYEARS";
    DateRangeType["LASTYEAR"] = "LASTYEAR";
    DateRangeType["YEAR"] = "YEAR";
    DateRangeType["NEXTDAYS"] = "NEXTDAYS";
    DateRangeType["NEXTWEEKS"] = "NEXTWEEKS";
    DateRangeType["NEXTMONTHS"] = "NEXTMONTHS";
    DateRangeType["NEXTQUARTERS"] = "NEXTQUARTERS";
    DateRangeType["NEXTYEARS"] = "NEXTYEARS";
    DateRangeType["NEXT"] = "NEXT";
    DateRangeType["SPECIFICMONTH"] = "SPECIFICMONTH";
    DateRangeType["YESTERDAY"] = "YESTERDAY";
    DateRangeType["YEARTODATE"] = "YEARTODATE";
    DateRangeType["TODAY"] = "TODAY";
    DateRangeType["TOMORROW"] = "TOMORROW";
    DateRangeType["THISWEEK"] = "THISWEEK";
    DateRangeType["LASTWEEK"] = "LASTWEEK";
    DateRangeType["LAST2WEEKS"] = "LAST2WEEKS";
    DateRangeType["LAST3WEEKS"] = "LAST3WEEKS";
    DateRangeType["LAST4WEEKS"] = "LAST4WEEKS";
    DateRangeType["LAST5WEEKS"] = "LAST5WEEKS";
    DateRangeType["NEXTWEEK"] = "NEXTWEEK";
    DateRangeType["NEXT2WEEKS"] = "NEXT2WEEKS";
    DateRangeType["NEXT3WEEKS"] = "NEXT3WEEKS";
    DateRangeType["NEXT4WEEKS"] = "NEXT4WEEKS";
    DateRangeType["NEXT5WEEKS"] = "NEXT5WEEKS";
    DateRangeType["THISMONTH"] = "THISMONTH";
    DateRangeType["LASTMONTH"] = "LASTMONTH";
    DateRangeType["NEXTMONTH"] = "NEXTMONTH";
    DateRangeType["THISQUARTER"] = "THISQUARTER";
    DateRangeType["LASTQUARTER"] = "LASTQUARTER";
    DateRangeType["NEXTQUARTER"] = "NEXTQUARTER";
    DateRangeType["QUARTER1"] = "QUARTER1";
    DateRangeType["QUARTER2"] = "QUARTER2";
    DateRangeType["QUARTER3"] = "QUARTER3";
    DateRangeType["QUARTER4"] = "QUARTER4";
    DateRangeType["TODAYFROMTO"] = "TODAYFROMTO";
})(DateRangeType = exports.DateRangeType || (exports.DateRangeType = {}));
