// @sap/ux-specification/dist/src/sync/v2/import/pages/analyticalListPage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../common/index");
const common_1 = require("../../../../specification/common");
const extensionLogger_1 = require("../../../../extensionLogger");
const i18n_1 = require("../../../../i18n/i18n");
const i18next_1 = __importDefault(require("i18next"));
const common_2 = require("../../../common");
const utils_1 = require("../../import/utils");
const factory_1 = require("../../export/factory");
/**
 * Adds manifest settings to the config.json
 * @param {AnalyticalListPageConfig} analyticalListPageConfig - the configuration to be filled
 * @param {Manifest} manifest - manifest.json of the app
 * @param {object} jsonSchema - the app-specific schema
 * @param {MetadataInstanceInterface} factory - factory, for generating the access to reflect-metadata
 * @param {string} pageKey - key of the page in manifest
 * @param {ExtensionLogger} logger - instance for logging exceptions
 */
function addSettings(importParameters, analyticalListPageConfig, factory, pageKey) {
    const tableType = index_1.determineTableType(importParameters.manifest, pageKey);
    analyticalListPageConfig.table = factory.createInstance(common_1.PageType.AnalyticalListPage, `ALP${tableType}`, analyticalListPageConfig.table);
    index_1.transferSettingsOfObject(analyticalListPageConfig.table, importParameters.manifest, importParameters.jsonSchema['definitions'][`ALP${tableType}<LineItems>`], pageKey, importParameters.logger);
    // ToolBar:
    analyticalListPageConfig.table.toolBar = factory.createInstance(common_1.PageType.AnalyticalListPage, 'ToolBar', {
        actions: {}
    });
    index_1.transferSettingsOfObject(analyticalListPageConfig.table.toolBar, importParameters.manifest, importParameters.jsonSchema['definitions']['ToolBar'], pageKey, importParameters.logger);
    if (importParameters.jsonSchema['definitions']['Actions<LineItems>']) {
        for (const actionProperty in importParameters.jsonSchema['definitions']['Actions<LineItems>'].properties) {
            const actionKey = actionProperty.replace('/', '::');
            analyticalListPageConfig.table.toolBar.actions[actionKey] = factory.createInstance(common_1.PageType.AnalyticalListPage, 'Action');
            index_1.transferSettingsOfObject(analyticalListPageConfig.table.toolBar.actions[actionKey], importParameters.manifest, importParameters.jsonSchema['definitions']['Action'], pageKey, importParameters.logger);
        }
    }
    //FilterBar:
    analyticalListPageConfig.filterBar = factory.createInstance(common_1.PageType.AnalyticalListPage, common_2.DefinitionName.AnalyticalListPageFilterBar, analyticalListPageConfig.filterBar);
    index_1.transferSettingsOfObject(analyticalListPageConfig.filterBar, importParameters.manifest, importParameters.jsonSchema['definitions'][common_2.DefinitionName.AnalyticalListPageFilterBar], pageKey, importParameters.logger);
    if (importParameters.jsonSchema['definitions']['SelectionFields']) {
        analyticalListPageConfig.filterBar.selectionFields = factory.createInstance(common_1.PageType.AnalyticalListPage, common_2.DefinitionName.AnalyticalListPageFilterBar, analyticalListPageConfig.filterBar.selectionFields);
        const schemaSelectionFields = importParameters.jsonSchema['definitions']['SelectionFields'].properties;
        for (const selectionProperty in schemaSelectionFields) {
            const selectionKey = selectionProperty;
            analyticalListPageConfig.filterBar.selectionFields[selectionKey] = factory.createInstance(common_1.PageType.AnalyticalListPage, 'SelectionFieldDateSettings');
            index_1.transferSettingsOfObject(analyticalListPageConfig.filterBar.selectionFields[selectionKey], importParameters.manifest, importParameters.jsonSchema['definitions']['SelectionFieldDateSettings'], pageKey, importParameters.logger, selectionKey);
        }
    }
    //Chart settings:
    analyticalListPageConfig.chart = factory.createInstance(common_1.PageType.AnalyticalListPage, 'ChartSettings', analyticalListPageConfig.chart);
    index_1.transferSettingsOfObject(analyticalListPageConfig.chart, importParameters.manifest, importParameters.jsonSchema['definitions']['ChartSettings'], pageKey, importParameters.logger);
    //Footer:
    analyticalListPageConfig.footer = { actions: {} };
    const footerAction = importParameters.jsonSchema['definitions']['FooterActions<LineItems>'];
    if (footerAction) {
        for (const actionProperty in footerAction.properties) {
            const actionKey = actionProperty.replace('/', '::');
            analyticalListPageConfig.footer.actions[actionKey] = factory.createInstance(common_1.PageType.AnalyticalListPage, 'FooterAction');
            index_1.transferSettingsOfObject(analyticalListPageConfig.footer.actions[actionKey], importParameters.manifest, importParameters.jsonSchema['definitions']['Action'], pageKey, importParameters.logger);
        }
    }
    //KPI's:
    analyticalListPageConfig.keyPerformanceIndicators = {};
    const kpiCards = importParameters.manifest[common_1.ManifestSection.generic]['pages'][pageKey].component.settings
        .keyPerformanceIndicators;
    if (kpiCards) {
        Object.keys(kpiCards).forEach((kpiKey) => {
            analyticalListPageConfig.keyPerformanceIndicators[kpiKey] = factory.createInstance(common_1.PageType.AnalyticalListPage, 'KPISettings');
            analyticalListPageConfig.keyPerformanceIndicators = factory.createInstance(common_1.PageType.AnalyticalListPage, 'KPISettings', analyticalListPageConfig.keyPerformanceIndicators);
            index_1.transferSettingsOfObject(analyticalListPageConfig.keyPerformanceIndicators[kpiKey], importParameters.manifest, importParameters.jsonSchema['definitions']['KPISettings'], pageKey, importParameters.logger, kpiKey);
        });
    }
}
/**
 * Creates the configuration file content for an analytical list page V2
 * @param {ImportALPV2Parameters} importParameters  - object comprising all input data
 *
 * @returns {AnalyticalListPageConfig} - the configuration (JSON)
 */
function createAnalyticalListPageConfig(importParameters) {
    // Initialize i18next
    i18n_1.initI18n();
    if (!importParameters.manifest[common_1.ManifestSection.generic]) {
        extensionLogger_1.log(importParameters.logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOFE'),
            location: {
                path: common_2.MANIFESTPATH,
                range: [common_1.ManifestSection.generic]
            }
        });
        return;
    }
    const v2Pages = importParameters.manifest[common_1.ManifestSection.generic].pages;
    if (!v2Pages) {
        extensionLogger_1.log(importParameters.logger, {
            severity: "error" /* Error */,
            message: i18next_1.default.t('NOPAGES', { appId: importParameters.manifest['sap.app']['id'] }),
            location: {
                path: common_2.MANIFESTPATH,
                range: [common_1.ManifestSection.generic]
            }
        });
        return;
    }
    const factory = new factory_1.MetadataInstanceFactoryV2();
    //Instantiate Page
    const config = factory.createPageInstance(common_1.PageType.AnalyticalListPage);
    const pageKeys = [];
    const v2Page = utils_1.findAnalyticalListPage(importParameters.manifest[common_1.ManifestSection.generic].pages, pageKeys, importParameters.logger);
    if (!v2Page) {
        return;
    }
    addSettings(importParameters, config, factory, pageKeys[0]);
    importParameters.flex.forEach((change) => {
        if (common_2.parseChangeString(change)) {
            index_1.addFlex(config, change, importParameters.jsonSchema);
        }
    });
    index_1.addFragments(config, importParameters.manifest, v2Page, common_1.PageType.AnalyticalListPage, importParameters.fragments);
    utils_1.removeEmptySettings(config);
    return config;
}
exports.createAnalyticalListPageConfig = createAnalyticalListPageConfig;
