// @sap/cds-mtx/lib/jobs/job.js
const GlobalConnections = require('../helper/global_connections');

const cds = require('@sap/cds');

const EXECUTION_STATUS = {
    INITIAL: 'INITIAL',
    QUEUED: 'QUEUED',
    RUNNING: 'RUNNING',
    FINISHED: 'FINISHED',
    FAILED: 'FAILED'
};

class Job {
    constructor(id, tenantId, context, jobFunction, args) {
        this.tenantId = tenantId;
        this.jobFunction = jobFunction;
        this.functionArgs = args;
        this.functionStatus = {};
        this.jobStatus = EXECUTION_STATUS.QUEUED;
        this.jobId = id;
        this.result = null;
        this.processId = process.pid;
        this.context = context;
    }

    static get EXECUTION_STATUS() {
        return EXECUTION_STATUS;
    }

    async setJobStatus(status) {
        this.jobStatus = status;
        await GlobalConnections.applicationDataPersistence.upsertJobStatus(this);
    }

    get error() {
        return this.jobStatus === EXECUTION_STATUS.FAILED ? this.result : null;
    }

    async run() {
        cds.context = this.context;
        try {
            this.result = await this.jobFunction(this.functionStatus, ...this.functionArgs);
            await this.setJobStatus(EXECUTION_STATUS.FINISHED);
        } catch (error) {
            this.result = error;
            await this.setJobStatus(EXECUTION_STATUS.FAILED);
            throw error;
        }

        return this.result;
    }
}

module.exports = Job;
