// @sap/approuter/lib/utils/header-util.js
'use strict';

const SAPPassport = require('@sap/e2e-trace').Passport;
const uuid = require('uuid').v4;

module.exports = {

  NOCACHE_HEADER_VALUE: 'no-cache, no-store, must-revalidate',

  updateSapPassport: function (headers, optTargetHeaders) {
    if (process.env.SERVICE_2_APPROUTER === 'true') {
      return;
    }
    const passportHeader = headers[SAPPassport.HEADER_NAME];
    if (passportHeader) {
      const passport = new SAPPassport(passportHeader);
      passport.update({
        previousComponent: 'XSA Approuter',
        connectionID: '00000000000000000000000000000000',
        connectionCounter: 0
      });
      const headersToUpdate = optTargetHeaders || headers;
      headersToUpdate[SAPPassport.HEADER_NAME] = passport.serialize();
    }
  },
  addCorrelationIdHeader: function (req) {
    if (req.headers['x-correlationid']) {
      return;
    } else if (req.headers['x-request-id']) {
      req.headers['x-correlationid'] = req.headers['x-request-id'];
    } else if (req.headers['x-vcap-request-id']) {
      req.headers['x-correlationid'] = req.headers['x-vcap-request-id'];
    } else {
      req.headers['x-correlationid'] = uuid();
    }
  },
  getCorrelationId: function (req) {
    let correlationId = (req && req.headers) ? req.headers['x-correlationid'] : uuid();
    return correlationId;
  }
};
