// @sap/ux-specification/dist/src/sync/v2/export/controls/ObjectPageSectionsV2.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const common_1 = require("../../../common");
exports.FacetTitlePrefix = 'Facet ID: ';
exports.getTableSectionControlId = (baseId, idBreadcrumbs, breadcrumbs, controlType, facetId) => {
    const convertedId = common_1.convertSectionId(breadcrumbs[1], facetId);
    const breadcrumbsTrail = [convertedId, common_1.DefinitionName.Table];
    return common_1.buildControlIdFromParent(baseId, breadcrumbsTrail);
};
exports.getFormSectionControlId = (baseId, idBreadcrumbs, breadcrumbs, controlType, facetId) => {
    const convertedId = common_1.convertSectionId(breadcrumbs[1], facetId);
    const breadcrumbsTrail = [convertedId, 'FormGroup'];
    return common_1.buildControlIdFromParent(baseId, breadcrumbsTrail);
};
exports.getChartSectionControlId = (baseId, idBreadcrumbs, breadcrumbs, controlType, facetID) => {
    const convertedId = common_1.convertSectionId(breadcrumbs[1], facetID);
    const breadcrumbsTrail = [convertedId, 'Chart'];
    return common_1.buildControlIdFromParent(baseId, breadcrumbsTrail);
};
exports.getSectionControlId = (baseId, idBreadcrumbs, breadcrumbs, controlType, facetId) => {
    const convertedId = common_1.convertSectionId(breadcrumbs[1], facetId);
    const breadcrumbsTrail = [convertedId, 'Section'];
    return common_1.buildControlIdFromParent(baseId, breadcrumbsTrail);
};
exports.getSubSectionControlId = (baseId, idBreadcrumbs, breadcrumbs, controlType, facetId) => {
    const convertedId = common_1.convertSectionId(breadcrumbs[1], facetId);
    const breadcrumbsTrail = [convertedId, 'SubSection'];
    return common_1.buildControlIdFromParent(baseId, breadcrumbsTrail);
};
class ObjectPageSectionV2 {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => common_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionV2.prototype, "visible", void 0);
exports.ObjectPageSectionV2 = ObjectPageSectionV2;
class ObjectPageSectionFormV2 {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => common_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionFormV2.prototype, "visible", void 0);
exports.ObjectPageSectionFormV2 = ObjectPageSectionFormV2;
class ObjectPageSectionChartV2 {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => common_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionChartV2.prototype, "visible", void 0);
exports.ObjectPageSectionChartV2 = ObjectPageSectionChartV2;
class ObjectPageSectionTableV2 {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: exports.getSectionControlId,
            controlType: () => common_1.ControlType.Section,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSectionTableV2.prototype, "visible", void 0);
exports.ObjectPageSectionTableV2 = ObjectPageSectionTableV2;
class ObjectPageSubSectionV2 {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: exports.getSubSectionControlId,
            controlType: () => common_1.ControlType.SubSection,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionV2 = ObjectPageSubSectionV2;
class ObjectPageSubSectionFormV2 {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: exports.getFormSectionControlId,
            controlType: () => common_1.ControlType.SmartGroup,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionFormV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionFormV2 = ObjectPageSubSectionFormV2;
class ObjectPageSubSectionChartV2 {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: exports.getChartSectionControlId,
            controlType: () => common_1.ControlType.SmartChart,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionChartV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionChartV2 = ObjectPageSubSectionChartV2;
class ObjectPageSubSectionTableV2 {
}
__decorate([
    decorators_1.syncRule({
        flex: {
            controlId: exports.getTableSectionControlId,
            controlType: () => common_1.ControlType.SmartTable,
            createFlexChange: common_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageSubSectionTableV2.prototype, "visible", void 0);
exports.ObjectPageSubSectionTableV2 = ObjectPageSubSectionTableV2;
