// @sap/ux-specification/dist/src/extensionLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var LogSeverity;
(function (LogSeverity) {
    LogSeverity["Error"] = "error";
    LogSeverity["Warning"] = "warning";
    LogSeverity["Info"] = "info";
})(LogSeverity = exports.LogSeverity || (exports.LogSeverity = {}));
exports.LogSeverityLabel = {
    ["error" /* Error */]: 'Error',
    ["warning" /* Warning */]: 'Warning',
    ["info" /* Info */]: 'Information'
};
/**
 * Function to create a new log entry and add it to the logger.
 * If no logger is passed, puts the message to the console
 * @param logger - the logger instance
 * @param message - the log message
 * @param meta - option list of strings/objects
 */
exports.log = (logger, logEntry, ...meta) => {
    if (logger) {
        logger[logEntry.severity](logEntry.message, logEntry.location, ...meta);
    }
    else {
        if (logEntry.location) {
            if (logEntry.location.range) {
                console.log(`${exports.LogSeverityLabel[logEntry.severity]}: ${logEntry.message}, location: ${logEntry.location.path}, ${logEntry.location.range}`, ...meta);
            }
            else {
                console.log(`${exports.LogSeverityLabel[logEntry.severity]}: ${logEntry.message}, location: ${logEntry.location.path}`, ...meta);
            }
        }
        else {
            console.log(`${exports.LogSeverityLabel[logEntry.severity]}: ${logEntry.message}`, ...meta);
        }
    }
};
