// @sap/ux-specification/dist/src/sync/v4/export/pages/ListReport.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
/**
 * sync rule deleting all views if the chart view gets deleted
 * @param manifestSection - manifest section of the given view
 */
function deleteAllALPViews(manifestSection) {
    var _a, _b, _c, _d;
    if (((_b = (_a = manifestSection['paths']) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.primary) || ((_d = (_c = manifestSection['paths']) === null || _c === void 0 ? void 0 : _c[0]) === null || _d === void 0 ? void 0 : _d.secondary)) {
        Object.keys(manifestSection).forEach((key) => delete manifestSection[key]);
    }
}
/**
 * sync rule for preventing the export of a property
 */
function doNotExport() {
    return;
}
/**
 * sync rule for preventing the import of a property
 */
function doNotImport() {
    return;
}
/**
 * Sync class for import and export of V4 list report
 */
let ListReport = class ListReport {
};
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet) => application_1.getSettingsPathForPage(entitySet) + '/views',
            import: doNotImport,
            export: doNotExport,
            delete: deleteAllALPViews
        }
    })
], ListReport.prototype, "chart", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet) => application_1.getSettingsPathForPage(entitySet)
        }
    }),
    decorators_1.validity({
        since: '1.84.0'
    })
], ListReport.prototype, "variantManagement", void 0);
__decorate([
    decorators_1.syncRule({
        manifest: {
            path: (entitySet) => application_1.getSettingsPathForPage(entitySet) + '/views/paths/[]/0'
        }
    })
], ListReport.prototype, "defaultPath", void 0);
__decorate([
    decorators_1.childId('fe::FilterBar')
], ListReport.prototype, "filterbar", void 0);
ListReport = __decorate([
    decorators_1.pageLayoutInformation({
        id: 'sap.suite.ui.generic.template.ListReport.view.ListReport',
        pageId: 'sap.suite.ui.generic.template.ListReport'
    })
], ListReport);
exports.ListReport = ListReport;
