// @sap/cds-mtx/lib/helper/cf/instance_manager_factory.js
const { promisify } = require('util');
const createInstanceManager = promisify(require('@sap/instance-manager').create);
const migrationSupport = require('@sap/instance-manager').hasMigrationSupport;
const { support, ERRORS } = require('../../../config/supportability');
const Logger = require('../../helper/logger');

let _instanceManager;

class InstanceManagerFactory {

    static _getServiceCredentials() {
        const cds = require('../../cds');

        if (!cds.env.requires.db) {
            support(ERRORS.hanaServiceMissing, InstanceManagerFactory.logger);
            throw new Error('HANA credentials could not be found.');
        }

        if (!cds.env.requires.db.credentials) {
            support(ERRORS.hanaCredentialsMissing, InstanceManagerFactory.logger);
            throw new Error('HANA credentials could not be found.');
        }

        const credentials = cds.env.requires.db.credentials;

        // add instance manager credentials if available
        // explicitly using vcap directly. This is only needed for container migration on cf
        if (cds.env.features.hybrid_instance_manager && process.env.VCAP_SERVICES) {
            const vcap = JSON.parse(process.env.VCAP_SERVICES)
            if (vcap['managed-hana'] && vcap['service-manager']) {

                if (!migrationSupport) {
                    throw new Error('Used version of @sap/instance-manager does not support binding of both managed-hana and instance-manager');
                }

                return {
                    smOpts: vcap['service-manager'][0].credentials,
                    imOpts: vcap['managed-hana'][0].credentials
                }
            }
        }

        return credentials;
    }

    static get logger() {
        return this._logger = this._logger || Logger('INSTANCE_MANAGER_FACTORY');
    }

    static async _create() {
        let options;
        try {
            options = InstanceManagerFactory._getServiceCredentials();
        } catch (error) {
            InstanceManagerFactory.logger.error(error);
            throw new Error(`Failed to get instance manager service credentials: ${error.message}`);
        }

        options = Object.assign(options, {
            cache_max_items: 1,
            cache_item_expire_seconds: 1,
            polling_timeout_seconds: 240,
        });

        const instanceManager = await createInstanceManager(options);
        instanceManager.create = promisify(instanceManager.create, instanceManager);
        instanceManager.delete = promisify(instanceManager.delete, instanceManager);
        instanceManager.get = promisify(instanceManager.get, instanceManager);
        instanceManager.getAll = promisify(instanceManager.getAll, instanceManager);
        _instanceManager = instanceManager;
    }

    static async get() {
        if (!_instanceManager) {
            await this._create();
        }
        return _instanceManager;
    }
}

module.exports = InstanceManagerFactory;
