// @sap/cds-mtx/config/supportability.js
//const { PERMALINKS } = require('./constants');

function support(error, logger) {
    const { diagnosis, support, permalink } = error;
    const capireReference = permalink ? `Find more information here: ${permalink}` : "";
    const message = [diagnosis, support, capireReference].filter(c => !!c).join("\n");
    logger.error(message);
}
const ERRORS = Object.freeze({
    hanaCredentialsMissing: {
        diagnosis: `Managed HANA service credentials not found.`,
        support: `Deploy your HANA database using 'cds deploy --to hana.'`,
        //permalink: PERMALINKS.CDS_DEPLOY_HANA,
    },
    hanaServiceMissing: {
        diagnosis: `HANA service not provided in CDS requirements.`,
        support: `Add 'hana' to 'cds.requires.db' in your package.json or .cdsrc.json.'`,
        //permalink: PERMALINKS.DEPLOY_AS_MULTITENANT,
    },
    uaaEnvRequiresMissing: {
        diagnosis: `The UAA credentials couldn't be found as no 'requires' are specified in the CDS environment configuration.`,
        //permalink: PERMALINKS.MULTITENANCY_CONFIGURATION,
    },
    uaaBindingMissing: {
        diagnosis: `The UAA credentials couldn't be found as the 'uaa' requirement is missing in the CDS configuration.`,
        //permalink: PERMALINKS.MULTITENANCY_CONFIGURATION,
    },
    uaaBindingNotXSUAA: {
        diagnosis: `The UAA credentials couldn't be read as the 'uaa' requirement in the CDS configuration has to be of type 'xsuaa'.`,
        //permalink: PERMALINKS.MULTITENANCY_CONFIGURATION,
    }
})

module.exports = { support, ERRORS }
