// @sap/ux-specification/dist/src/sync/v2/export/factory.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const v2_1 = require("../../../specification/v2");
const controls = __importStar(require("./controls"));
const pages = __importStar(require("./pages"));
const factory_1 = require("../../common/decoration/factory");
const common_1 = require("../../../specification/common");
/**
 * Note: If the key of the elements corresponds to the name of the definition in schema, you do not have to list it explicitly here.
 * The following tables only comprise entries for elements that do not follow the official pattern (yet)
 */
const ALPElements = {
//no deviations
};
const LRElements = {
    Table: controls.ResponsiveTable,
    ResponsiveTableWithMultiSelect: controls.ResponsiveTable,
    ResponsiveTableWithInlineDelete: controls.ResponsiveTable
};
const OVPElements = {
    [v2_1.CardSettingsType.listCardSettings]: controls.ListCard,
    [v2_1.CardSettingsType.listCardSettingsv4]: controls.ListCard,
    [v2_1.CardSettingsType.analyticalCardSettings]: controls.AnalyticalCard,
    [v2_1.CardSettingsType.analyticalCardSettingsv4]: controls.AnalyticalCard,
    [v2_1.CardSettingsType.stackCardSettings]: controls.StackCard,
    [v2_1.CardSettingsType.linkListCardSettings]: controls.LinklistCard,
    [v2_1.CardSettingsType.tableCardSettings]: controls.TableCard,
    [v2_1.CardSettingsType.tableCardSettingsv4]: controls.TableCard
};
const OPElements = {
    [common_1.SAPUI5_VIEW_CLASS]: controls.ObjectPageCustomSectionView,
    [common_1.SAPUI5_FRAGMENT_CLASS]: controls.ObjectPageCustomSectionFragment,
    Table: controls.ObjectPageResponsiveTable,
    ObjectPageResponsiveTableWithMultiSelect: controls.ObjectPageResponsiveTable,
    ObjectPageResponsiveTableWithInlineDelete: controls.ObjectPageResponsiveTable
};
const pageTypes = {
    [common_1.PageType.AnalyticalListPage]: ALPElements,
    [common_1.PageType.ListReport]: LRElements,
    [common_1.PageType.ObjectPage]: OPElements,
    [common_1.PageType.OverviewPage]: OVPElements
};
/**
 * V2 specific factory class for creating instances of (reflect) metadata classes
 */
class MetadataInstanceFactoryV2 extends factory_1.MetadataFactory {
    constructor() {
        super(controls, pages, pageTypes);
    }
}
exports.MetadataInstanceFactoryV2 = MetadataInstanceFactoryV2;
